File: 01_GOF23/DP_examples/Go/02_Structural/Adapter/adaptee.go
package Adapter

import "fmt"

// 媒体类型枚举
type MediaType int

const (
	MP3 MediaType = iota
	AIFF
	MIDI
	Wave
	MP4
	MOV
	AVI
	M4V
)

// 被适配的接口
type IAdvancedMediaPlayer interface {
	PlayMedia(fileName string, typ MediaType)
}

// 音频播放器（被适配的类）
type AudioPlayer struct{}

func (ap *AudioPlayer) PlayMedia(fileName string, typ MediaType) {
	switch typ {
	case AIFF:
		fmt.Println("PlayAudio AIFF :", fileName)
	case MIDI:
		fmt.Println("PlayAudio MIDI :", fileName)
	case Wave:
		fmt.Println("PlayAudio wave :", fileName)
	}
}

// 视频播放器（被适配的类）
type VideoPlayer struct{}

func (vp *VideoPlayer) PlayMedia(fileName string, typ MediaType) {
	switch typ {
	case MP4:
		fmt.Println("PlayVideo mp4 :", fileName)
	case MOV:
		fmt.Println("PlayVideo mov :", fileName)
	case AVI:
		fmt.Println("PlayVideo avi :", fileName)
	case M4V:
		fmt.Println("PlayVideo m4v :", fileName)
	}
}

File: 01_GOF23/DP_examples/Rust/02_Structural/Adapter/adaptee.rs
// 媒体类型枚举
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum MediaType {
    MP3,
    AIFF,
    MIDI,
    Wave,
    MP4,
    MOV,
    AVI,
    M4V,
}

// 被适配的接口
pub trait IAdvancedMediaPlayer {
    fn play_media(&self, file_name: &str, typ: MediaType);
}

// 音频播放器（被适配的类）
pub struct AudioPlayer;

impl IAdvancedMediaPlayer for AudioPlayer {
    fn play_media(&self, file_name: &str, typ: MediaType) {
        match typ {
            MediaType::AIFF => println!("PlayAudio AIFF : {}", file_name),
            MediaType::MIDI => println!("PlayAudio MIDI : {}", file_name),
            MediaType::Wave => println!("PlayAudio wave : {}", file_name),
            _ => (),
        }
    }
}

// 视频播放器（被适配的类）
pub struct VideoPlayer;

impl IAdvancedMediaPlayer for VideoPlayer {
    fn play_media(&self, file_name: &str, typ: MediaType) {
        match typ {
            MediaType::MP4 => println!("PlayVideo mp4 : {}", file_name),
            MediaType::MOV => println!("PlayVideo mov : {}", file_name),
            MediaType::AVI => println!("PlayVideo avi : {}", file_name),
            MediaType::M4V => println!("PlayVideo m4v : {}", file_name),
            _ => (),
        }
    }
}
